### ISSUE TITLE
`plan`/`apply` raises 404 on `prefect_work_pool` if work pool is deleted outside Terraform

### ISSUE THREAD

**Issue by kevingrismore** [PrefectHQ/terraform-provider-prefect#260]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Terraform Version
```
Terraform v1.9.5
on darwin_arm64
+ provider registry.terraform.io/prefecthq/prefect v2.2.2
```

### Affected Resource(s)

* prefect_work_pool

### Terraform Configuration Files

```tf
resource "prefect_work_pool" "example" {
  name              = "test-k8s-pool"
  type              = "kubernetes"
  workspace_id      = "<workspace-id>"
  paused            = false
  base_job_template = file("./base-job-template.json")
}
```

### Expected Behavior

After creating a work pool with Terraform, then deleting it through the Prefect UI, Terraform should detect that the work pool doesn't exist & try to recreate it.

### Actual Behavior

After creating a work pool with Terraform, then deleting it through the Prefect UI, Terraform raises a 404.

```
│ Error: Error during get Work Pool
│ 
│   with prefect_work_pool.example,
│   on work_pool.tf line 3, in resource "prefect_work_pool" "example":
│    3: resource "prefect_work_pool" "example" {
│ 
│ Could not get Work Pool, unexpected error: status code 404 Not Found, error={"detail":"Work pool \"test-k8s-pool\" not found."}
```

### Steps to Reproduce

1. Define a work pool resource
2. `terraform apply`
3. Delete the work pool from the Prefect UI
4. `terraform apply`

**Comment by mitchnielsen** [comment]
**TL;DR**: from what I've seen so far, this is expected behavior and you can use `terraform state rm` to remove the Work Pool from Terraform's state, and then run the plan again to recreate the resource. If anyone has further evidence to the contrary, or if we get enough requests, I'm completely open to revisiting this.

-----

Did some digging on this one. From what I've seen so far, this is somewhat expected behavior (out of the box, at least).

You'd typically use [terraform state rm](https://developer.hashicorp.com/terraform/cli/commands/state/rm) if you notice that a resource has been deleted outside of Terraform. That would allow it to correctly create the resource during the next run - I tested this to confirm:

1. Created a work pool via Terraform.
2. Deleted the work pool in the UI.
3. Ran the Terraform plan again, and got the 404 error.
4. Ran `terraform state rm prefect_work_pool.test_work_pool`.
5. Ran the Terraform plan again and it correctly recreated the pool.

That said, I have seen a [reference](https://discuss.hashicorp.com/t/recreate-a-resource-in-a-case-of-manual-deletion/66375/2) that suggests you can add some extra logic in the `Read` methods where you mark the object as `null` if it's not found, and that would also let Terraform recreate the resource:

> ... the way to handle that is for the Read operation to return a null object if it finds that the object is missing, because null is how Terraform represents “does not exist”. The entire object must be null to represent not existing; setting just the id to null isn’t sufficient because Terraform Core doesn’t treat the “id” attribute in any special way here.

> If you make your Read behave in that way then the built-in behavior of the plugin framework should handle the planning step for you: it will notice that the prior state is null but the desired state isn’t, and so will return something derived from the desired state as the planned new state, and Terraform Core will record that as a “create” action, which you can then handle during the apply phase as normal.

That sounds somewhat reasonable, although I am curious how you differentiate between "I didn't find the object because it didn't exist" and "I didn't find the object because of any number of transient problems" - maybe our API gives a consistently unique error code and we could work around that.

I haven't yet found this as a documented / recommended part of the Terraform provider guidance other than forum posts, so I'm going to lean toward avoiding this extra logic for now. If I (or anyone else) finds that we already do this somewhere or that there's a reliable way to implement it, I'm definitely open to revisiting this.

**Comment by murukesh-mohanan-paidy** [comment]
@mitchnielsen I have to disagree on running `terraform state rm` being "typical" to handle the case where a resource vanished outside of Terraform. In my day-to-day work, the popular providers that I use (Hashicorp-maintained ones like the AWS or GCP providers, the GitHub provider, the Datadog provider, etc. etc.) _all_ offer to create the resource anew, and update or recreate other resources that depend on the deleted resource. The forum post you link to is essentially a workaround when the provider doesn't handle this cleanly. I know the original post is about the AWS provider, but for run-of-the-mill resourced, the AWS provider does handle it just fine.

For example, you could have something like:

resource "aws_ssm_parameter" "test" {
  name  = "test"
  type  = "SecureString"
  value = "ignore"
}

resource "kubernetes_secret" "test" {
  metadata {
    name = "test"
  }
  data = {
    secret = aws_ssm_parameter.test.value
  }
}

**Comment by mitchnielsen** [comment]
Thanks for reaching out @murukesh-mohanan-paidy, we've actually started doing this but only for certain resources (so far). In https://github.com/PrefectHQ/terraform-provider-prefect/pull/423/files, we turned it on for work pools and work queues. We're happy to also try this for deployments and deployment schedules since that seems to be where you were noticing the problem. If you have any other top-priority resources, just let us know 👍🏼 

I've reopened https://github.com/PrefectHQ/terraform-provider-prefect/issues/417 to continue tracking this work.

**Comment by mitchnielsen** [comment]
This has been enabled for quite a few resources now:

<img width="225" alt="Image" src="https://github.com/user-attachments/assets/e57dbfc7-4b18-4407-a98a-fae0115d5a08" />

It's available in [v2.27.0](https://github.com/PrefectHQ/terraform-provider-prefect/releases/tag/v2.27.0) - reach out with any feedback!
